library/alloc/src/string.rs

```diff
@@ -2224,6 +2224,62 @@ impl ToString for char {
     }
 }
 
+#[stable(feature = "u8_to_string_specialization", since = "1.999.0")]
+impl ToString for u8 {
+    #[inline]
+    fn to_string(&self) -> String {
+        let mut result = String::with_capacity(3);
+        let mut n = *self;
+        if n >= 100 {
+            result.push((b'0' + n / 100) as char);
+            n %= 100;
+        }
+        if !result.is_empty() || n >= 10 {
+            result.push((b'0' + n / 10) as char);
+            n %= 10;
+        };
+        result.push((b'0' + n) as char);
+        result
+    }
+}
+
+// 2 digit decimal look up table
+static DEC_DIGITS_LUT: &[u8; 200] = b"0001020304050607080910111213141516171819\
+      2021222324252627282930313233343536373839\
+      4041424344454647484950515253545556575859\
+      6061626364656667686970717273747576777879\
+      8081828384858687888990919293949596979899";
+
+#[stable(feature = "i8_to_string_specialization", since = "1.999.0")]
+impl ToString for i8 {
+    #[inline]
+    fn to_string(&self) -> String {
+        let mut n = *self;
+        let mut vec: Vec<u8> = if n < 0 {
+            // convert the negative num to positive by summing 1 to it's 2 complement
+            // ( -128u8.abs() would panic )
+            n = (!n).wrapping_add(1);
+            let mut v = Vec::with_capacity(4);
+            v.push(b'-');
+            v
+        } else {
+            Vec::with_capacity(3)
+        };
+        let mut n = n as u8;
+        if n >= 10 {
+            if n >= 100 {
+                n -= 100;
+                vec.push(b'1');
+            }
+            let nn = n * 2;
+            vec.extend_from_slice(&DEC_DIGITS_LUT[nn as usize..=nn as usize + 1]);
```

Comment:
1. `n` can now be 128, so that multiplication will overflow. You need to cast first.
2. Unlike the cast to u8 above, these conversions are intended to be lossless. The type system can enforce that for you if you use `usize::from` instead of `as`. Also eases the mind of future readers.
3. You might want to check if using an exclusive range generates better code.

Reply:
Extending seems to make it hard for llvm to tell if the vec will grow. It turns out not using the lookup table is faster here.